ext/integrations/exec_integration.c

```diff
@@ -0,0 +1,365 @@
+
+    proc_h->npipes += 1;
+    proc_h->pipes = safe_erealloc(proc_h->pipes, proc_h->npipes, sizeof *proc_h->pipes, 0);
+    zend_resource *proc_span_res = zend_register_resource(proc_span, le_proc_span);
+    proc_h->pipes[proc_h->npipes - 1] = proc_span_res;
+
+    RETURN_TRUE;
+}
+
+static PHP_FUNCTION(DDTrace_integrations_exec_proc_get_span) {
+    zval *zres;
+
+    ZEND_PARSE_PARAMETERS_START(1, 1)
+    Z_PARAM_RESOURCE(zres)
+    ZEND_PARSE_PARAMETERS_END();
+
+    if (Z_RES_TYPE_P(zres) != le_proc) {
+        RETURN_NULL();
+    }
+
+    php_process_handle *proc_h = Z_RES_P(zres)->ptr;
+    if (proc_h->npipes == 0) {
+        RETURN_NULL();
+    }
+
+    zend_resource *span_res = proc_h->pipes[proc_h->npipes - 1];
+    if (span_res->type != le_proc_span) {
+        RETURN_NULL();
+    }
+
+    dd_proc_span *span_h = span_res->ptr;
+    RETURN_OBJ_COPY(span_h->span);
+}
+
+// used in testing only
+static PHP_FUNCTION(DDTrace_integrations_exec_proc_get_pid) {
+    zval *zres;
+
+    ZEND_PARSE_PARAMETERS_START(1, 1)
+    Z_PARAM_RESOURCE(zres)
+    ZEND_PARSE_PARAMETERS_END();
+
+    if (Z_RES_TYPE_P(zres) != le_proc) {
+        RETURN_NULL();
+    }
+
+    php_process_handle *proc_h = Z_RES_P(zres)->ptr;
+    RETURN_LONG((long)proc_h->child);
+}
+static PHP_FUNCTION(DDTrace_integrations_exec_test_rshutdown) {
+    if (zend_parse_parameters_none() != SUCCESS) {
+        return;
+    }
+
+    ddtrace_exec_handlers_rshutdown();
+    in_request = true;
+    (void)return_value;
+}
+
+// clang-format off
+ZEND_BEGIN_ARG_WITH_RETURN_TYPE_INFO_EX(register_stream, 0, 2, _IS_BOOL, 0)
+    ZEND_ARG_TYPE_INFO(0, stream, IS_RESOURCE, 0)
+    ZEND_ARG_OBJ_INFO(0, span, DDTrace\\SpanData, 0)
+ZEND_END_ARG_INFO()
+ZEND_BEGIN_ARG_WITH_RETURN_TYPE_INFO_EX(proc_assoc_span, 0, 2, _IS_BOOL, 1)
+    ZEND_ARG_TYPE_INFO(0, proc_res, IS_RESOURCE, 0)
+    ZEND_ARG_OBJ_INFO(0, span, DDTrace\\SpanData, 0)
+ZEND_END_ARG_INFO()
+ZEND_BEGIN_ARG_WITH_RETURN_OBJ_INFO_EX(proc_get_span, 0, 1, DDTrace\\SpanData, 0)
+    ZEND_ARG_TYPE_INFO(0, proc_res, IS_RESOURCE, 0)
+ZEND_END_ARG_INFO()
+ZEND_BEGIN_ARG_WITH_RETURN_TYPE_INFO_EX(proc_get_pid, 0, 1, IS_LONG, 1)
+    ZEND_ARG_TYPE_INFO(0, proc_res, IS_RESOURCE, 0)
+ZEND_END_ARG_INFO()
+ZEND_BEGIN_ARG_WITH_RETURN_TYPE_INFO_EX(test_rshutdown, 0, 0, IS_NULL, 1)
+ZEND_END_ARG_INFO()
+
+static const zend_function_entry functions[] = {
+    ZEND_RAW_FENTRY(NS "register_stream", PHP_FN(DDTrace_integrations_exec_register_stream), register_stream, 0)
+    ZEND_RAW_FENTRY(NS "proc_assoc_span", PHP_FN(DDTrace_integrations_exec_proc_assoc_span), proc_assoc_span, 0)
+    ZEND_RAW_FENTRY(NS "proc_get_span", PHP_FN(DDTrace_integrations_exec_proc_get_span), proc_get_span, 0)
+    ZEND_RAW_FENTRY(NS "proc_get_pid", PHP_FN(DDTrace_integrations_exec_proc_get_pid), proc_get_pid, 0)
+    ZEND_RAW_FENTRY(NS "test_rshutdown", PHP_FN(DDTrace_integrations_exec_test_rshutdown), test_rshutdown, 0)
+    PHP_FE_END
+};
```

Comment:
Can you please add the definitions to `ddtrace.stub.php`?
A pure testing function like `exec_test_reshutdown` should be proably hidden away in `dd_trace_internal_fn` - we're trying to be a bit tidy regarding testing functions :-)

Reply:
@cataphract You moved it to the stub file, but did not regenerate the arginfo: in php-src repo, there's a build/gen_stub.php script, which needs to be run.

Reply:
I put them in a different stub file.

As to moving them do `dd_trace_internal_fn` I don't think it's a good idea. I'd either have to duplicate logic from integrations_exec (fetching the resource type and a conditional typedef) or to export global functions in integrations_exec.c. Either way, it's a solution with less code cohesion. Also, proc_get_pid doesn't fit the signature of dd_trace_internal_fn, which returns a bool. The only disadvantage of the current solution is that are more PHP functions registered, but they are in a namespace that only has internal functions anyway.